op-supervisor: head pointers, refactor block processor

protolambda opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

Experimental draft of head block pointers instead of raw indices, and fixes of the x-head maintainer to use said block pointers.

Also includes an updated block-processor (the receipts fetcher), to be more asynchronous. I might split that out into a separate PR, TBD.

**Tests**

Work in progress.

axelKingsley commented on op-supervisor/supervisor/backend/source/chain.go:
```diff
@@ -43,26 +45,26 @@ func NewChainMonitor(ctx context.Context, logger log.Logger, m Metrics, chainID
-		log:         logger,
-		headMonitor: headMonitor,
+		log:            logger,
+		headMonitor:    headMonitor,
+		chainProcessor: unsafeBlockProcessor,
```
Not sure if this is intentional, but the `unsafeBlockProcessor` is now embedded twice:
* Once as this new `chainProcessor`
* Once inside `headMonitor` <- `callback` <- `unsafeProcessors`

Looks like this new `chainProcessor` value is just so we call `Close` on it? There's already `headMonitor.Stop` so I'm not sure the purpose of this.

protolambda replied:
Let's fix this in the follow-up PR

axelKingsley reviewed (commented):
Summary of changes as I understand them:

## Head Pointers
Rather than use EntryIdx to represent the pointer into the database, there's now an application specific structure which captures the Block Number and Index (called "Logs Since")

## Downstream Effects
* UpdateCrossHeadsForChain now tracks the x-head position using this pointer notion. The pointer can initialize itself in position without searching around. 

## Safety Checkers Return Err
Rather than return `bool` for the safety check, they now return an error related to which kind of issue is encountered (validity or inclusion). Safety Checkers also needed to be updated to use the new storage mechanisms

## Head Storage hides Update Functions
Now the storage has explicit functions to update the named safety levels, per chain. it handles making the update operation

## Chain processors Refactor
Seems like Chain Processors are now using a channel based trigger instead of the previous `OnNewHead`. I'm not sure if these changes are very connected to the rest of the PR. I think it's so that it doesn't block?
Additionally, the `headpointer_processor`s that I wrote yesterday are gone, in favor of using the explicit `heads` update functions, wrapped into satisfying the HeadProcessor interface.

axelKingsley commented on op-supervisor/supervisor/backend/db/db_test.go:
```diff
@@ -170,9 +171,9 @@ func setupStubbedForUpdateHeads(chainID types.ChainID) (*stubLogDB, *stubChecker
 	// the last known cross-safe head is at 20
-	cross := entrydb.EntryIdx(20)
+	cross := heads.HeadPointer{LastSealedBlockNum: 20}
 	// the local head (the limit of the update) is at 40
-	local := entrydb.EntryIdx(40)
+	local := heads.HeadPointer{LastSealedBlockNum: 40}
```
This change satisfied compilation errors for the tests but these tests are not meaningfully fixed yet.

protolambda committed:
op-supervisor: supervisor-head-pointers squashed

change entry indices to head pointers, refactor block processor, backend fixes

Co-authored-by: Axel Kingsley <axel.kingsley@gmail.com>

protolambda commented:
Ok, dropped the non-supervisor changes, rebased on develop, cleaned up the truncated-hash -> common-hash type change conflicts, and fixed some linting issues.
Tests are still in a very rough spot due to refactor and changed assumptions, we're going to leave them commented, and address in the follow-up PR here: https://github.com/ethereum-optimism/optimism/pull/12058

protolambda commented on op-supervisor/supervisor/backend/db/db_test.go:
```diff
@@ -213,39 +214,57 @@ func setupStubbedForUpdateHeads(chainID types.ChainID) (*stubLogDB, *stubChecker
+func (s *stubChecker) check(types.ChainID, uint64, uint32, backendTypes.TruncatedHash) error {
 	if s.checkCalls >= s.numSafe {
-		return false
+		return fmt.Errorf("safety check failed")
```
it's in commented tests, to be changed soon. Ignoring this.

axelKingsley committed:
use ticker instead of time.After
